### PR TITLE
Fixes SerializationService String serialization to obey the UTF-8 standard (RFC 3629) including the 4-byte characters

### PR DESCRIPTION
**Pull request by ihsandemir**
Fixes SerializationService 4-byte utf-8 support when serializing and de-serializing strings. The new format uses number of bytes for the length field instead of number of Java characters in the string. The fix enables string data type serialization to follow UTF-8 standard (RFC 3629).

The fix is for `Hazelcast Serialization Improvements` PRD.

### CODE REVIEW

**Comment by pveentjer** [comment]
Can you make a small JMH benchmark to see if we run into any performance regressions?

**Comment by puzpuzpuz** [comment]
> Can you make a small JMH benchmark to see if we run into any performance regressions?

A note on this topic. I had a JHM benchmark for string deserialization here:
https://github.com/puzpuzpuz/hazelcast-java-client-benchmarks/blob/master/src/main/java/com/hazelcast/StringDeserializationBenchmark.java

Maybe it'll save some time.

**Comment by pveentjer** [comment]
Can the results of the benchmark be posted on this PR?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataInput.java [hunk 1]
```diff
@@ -550,23 +551,15 @@ public int readUnsignedShort() throws EOFException {
...
-            b = readByte();
-            if (b < 0) {
-                charBuffer[i] = Bits.readUtf8Char(this, b);
-            } else {
-                charBuffer[i] = (char) b;
-            }
-        }
-        return new String(charBuffer, 0, charCount);
+
+        checkAvailable(pos, numberOfBytes);
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataInput.java]: It looks like the bounds check is also done by String constructor. Can you check if this line is still necessary ?

**Comment by puzpuzpuz** [comment]
> Can the results of the benchmark be posted on this PR?

Sure. Here they are (`length` means size in bytes; ASCII strings were used):
```
JDK 1.8.0_212, OpenJDK 64-Bit Server VM, 25.212-b03

Benchmark                                           (length)   Mode  Cnt         Score        Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom       100  thrpt    5  15071672.461 ± 882094.683  ops/s
StringDeserializationBenchmark.benchmarkStandard         100  thrpt    5  11236919.107 ± 836946.731  ops/s

Benchmark                                           (length)   Mode  Cnt        Score       Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom      1024  thrpt    5  1996938.463 ± 59222.347  ops/s
StringDeserializationBenchmark.benchmarkStandard        1024  thrpt    5  1880715.963 ± 36603.890  ops/s

Benchmark                                           (length)   Mode  Cnt      Score      Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom    102400  thrpt    5  19241.133 ± 1184.588  ops/s
StringDeserializationBenchmark.benchmarkStandard      102400  thrpt    5  24640.964 ± 2173.740  ops/s

JDK 11.0.3, OpenJDK 64-Bit Server VM, 11.0.3+7-Ubuntu-1ubuntu218.04.1

Benchmark                                           (length)   Mode  Cnt         Score         Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom       100  thrpt    5  14713769.433 ± 1001140.866  ops/s
StringDeserializationBenchmark.benchmarkStandard         100  thrpt    5  24617366.739 ±  255513.848  ops/s

Benchmark                                           (length)   Mode  Cnt         Score        Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom      1024  thrpt    5   1505573.119 ±  13399.126  ops/s
StringDeserializationBenchmark.benchmarkStandard        1024  thrpt    5  10301557.338 ± 265892.504  ops/s

Benchmark                                           (length)   Mode  Cnt       Score     Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom    102400  thrpt    5   12952.019 ± 379.904  ops/s
StringDeserializationBenchmark.benchmarkStandard      102400  thrpt    5  112547.060 ± 509.374  ops/s

JDK 12.0.1, Java HotSpot(TM) 64-Bit Server VM, 12.0.1+12

Benchmark                                           (length)   Mode  Cnt         Score         Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom       100  thrpt    5  14350468.240 ± 1043360.152  ops/s
StringDeserializationBenchmark.benchmarkStandard         100  thrpt    5  23692604.546 ±  218949.623  ops/s

Benchmark                                           (length)   Mode  Cnt         Score        Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom      1024  thrpt    5   1505765.159 ±  16182.719  ops/s
StringDeserializationBenchmark.benchmarkStandard        1024  thrpt    5  10130491.965 ± 169174.811  ops/s

Benchmark                                           (length)   Mode  Cnt       Score      Error  Units
StringDeserializationBenchmark.benchmarkIMDGCustom    102400  thrpt    5   12999.373 ±  552.527  ops/s
StringDeserializationBenchmark.benchmarkStandard      102400  thrpt    5  110778.502 ± 5848.281  ops/s
```

**Comment by puzpuzpuz** [comment]
A couple of questions:
1. I see that `Bits#readUtf8Char` is still being used by `DataInputNavigableJsonAdapter.UTF8Reader`. Could you check if it also has to be migrated to `new String()` API?
2. Does it make sense to completely remove `Bits#readUtf8Char` and `Bits#writeUtf8Char` methods to avoid confusion in future?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataInput.java (hunk 1 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/ByteArrayObjectDataInput.java]: Removed the line.

**Comment by ihsandemir** [comment]
I did run a similar benchmark as @puzpuzpuz and my finding are here as well: https://github.com/ihsandemir/utf8performance

**Comment by puzpuzpuz** [comment]
> A couple of questions:
> 
> 1. I see that `Bits#readUtf8Char` is still being used by `DataInputNavigableJsonAdapter.UTF8Reader`. Could you check if it also has to be migrated to `new String()` API?
> 2. Does it make sense to completely remove `Bits#readUtf8Char` and `Bits#writeUtf8Char` methods to avoid confusion in future?

@ihsandemir Could you check the above questions?

**Comment by ihsandemir** [comment]
Open an issue(https://github.com/hazelcast/hazelcast/issues/15402 ) for comment https://github.com/hazelcast/hazelcast/pull/15238#issuecomment-515875936

**Pull request merged**